File: src/crewlet/db/budgets.py
"""Shared token-budget usage: atomic spend across processes.

A budget splits into a cap and a counter, and only the counter is shared
state. The cap comes from the active config revision, so every process
derives the same number without coordinating. The counter answers "has
the company spent its 500k yet", which stops being true the moment it
lives in one process's memory: with N processes an org cap of 500k
becomes N x 500k, silently.

The spend is therefore a single conditional UPDATE per scope, with the
cap carried in the WHERE clause — check and increment in one statement,
so no peer can slip between them. Org and agent scopes move inside one
transaction, because a turn that is refused at the seat level must not
still have charged the org.

Both backends implement :class:`BudgetUsageStore` with identical
semantics and share one contract suite.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Protocol

from crewlet._logging import get_logger
from crewlet.db.client import Database

logger = get_logger("db.budgets")

ORG_SCOPE = "org"


def agent_scope(agent_id: str) -> str:
    """Usage scope for one seat."""
    return f"agent:{agent_id}"


@dataclass(frozen=True, slots=True)
class SpendOutcome:
    """The result of an attempted spend.

    ``rejected_scope`` names which budget refused (``"org"`` or
    ``"agent"``) so the caller can publish a ``BudgetExhausted`` that says
    which one, without a second read that a peer could have changed
    underneath it.
    """

    ok: bool
    org_used: int = 0
    agent_used: int = 0
    rejected_scope: str = ""
    rejected_used: int = 0
    rejected_limit: int = 0


class BudgetUsageStore(Protocol):
    """Atomic usage accounting for the org and one agent."""

    async def spend(
        self,
        *,
        agent_id: str,
        tokens: int,
        org_limit: int,
        agent_limit: int | None,
    ) -> SpendOutcome: ...

    async def usage(self, scope: str) -> int: ...

    async def reset(self, scope: str = "") -> int: ...


class _Refused(Exception):
    """Internal: unwinds the transaction when the second scope refuses."""

    def __init__(self, scope: str, used: int, limit: int) -> None:
        self.scope = scope
        self.used = used
        self.limit = limit


_BUMP_SQL = """
    INSERT INTO token_budget_usage (scope, used_tokens)
    VALUES ($1, $2)
    ON CONFLICT (scope) DO UPDATE
    SET used_tokens = token_budget_usage.used_tokens + $2,
        updated_at  = now()
    WHERE $3 = 0
       OR token_budget_usage.used_tokens + $2 <= $3
    RETURNING used_tokens
"""


class PostgresBudgetUsageStore:
    """PostgreSQL-backed :class:`BudgetUsageStore`."""

    def __init__(self, db: Database) -> None:
        self._db = db

    async def spend(
        self,
        *,
        agent_id: str,
        tokens: int,
        org_limit: int,
        agent_limit: int | None,
    ) -> SpendOutcome:
        if tokens <= 0:
            return SpendOutcome(ok=True)

        # A spend larger than a whole cap can never fit, and the INSERT
        # branch of the upsert (a scope with no row yet) has no existing
        # value to test against — so screen it here, before writing.
        for scope_name, limit in (("org", org_limit), ("agent", agent_limit)):
            if limit and tokens > limit:
                current = await self.usage(
                    ORG_SCOPE if scope_name == "org" else agent_scope(agent_id)
                )
                return SpendOutcome(
                    ok=False,
                    rejected_scope=scope_name,
                    rejected_used=current,
                    rejected_limit=limit,
                )

        try:
            async with self._db.transaction() as tx:
                org_row = await tx.fetchrow(
                    _BUMP_SQL, ORG_SCOPE, int(tokens), int(org_limit or 0)
                )
                if org_row is None:
                    raise _Refused(
                        "org", await self._usage_in(tx, ORG_SCOPE), int(org_limit or 0)
                    )
                org_used = int(org_row["used_tokens"])

                scope = agent_scope(agent_id)
                agent_row = await tx.fetchrow(
                    _BUMP_SQL, scope, int(tokens), int(agent_limit or 0)
                )
                if agent_row is None:
                    # Unwinds the org bump: the seat was refused, so the
                    # org must not be charged for a turn that never ran.
                    raise _Refused(
                        "agent", await self._usage_in(tx, scope), int(agent_limit or 0)
                    )
                return SpendOutcome(
                    ok=True,
                    org_used=org_used,
                    agent_used=int(agent_row["used_tokens"]),
                )
        except _Refused as refused:
            return SpendOutcome(
                ok=False,
                rejected_scope=refused.scope,
                rejected_used=refused.used,
                rejected_limit=refused.limit,
            )

    @staticmethod
    async def _usage_in(tx: Any, scope: str) -> int:
        row = await tx.fetchrow(
            "SELECT used_tokens FROM token_budget_usage WHERE scope = $1", scope
        )
        return int(row["used_tokens"]) if row else 0

    async def usage(self, scope: str) -> int:
        row = await self._db.fetchrow(
            "SELECT used_tokens FROM token_budget_usage WHERE scope = $1", scope
        )
        return int(row["used_tokens"]) if row else 0

    async def reset(self, scope: str = "") -> int:
        """Zero one scope, or every scope when ``scope`` is empty.

        Returns the number of scopes reset.
        """
        if scope:
            rows = await self._db.execute(
                "DELETE FROM token_budget_usage WHERE scope = $1 RETURNING scope",
                scope,
            )
        else:
            rows = await self._db.execute(
                "DELETE FROM token_budget_usage RETURNING scope"
            )
        return len(rows)


class MemoryBudgetUsageStore:
    """In-memory :class:`BudgetUsageStore` twin.

    Correct for a single process, which is what it is for. It does NOT
    make a multi-process deployment safe, and the engine refuses to use
    it as the authoritative counter when a database is configured.
    """

    def __init__(self) -> None:
        self._used: dict[str, int] = {}

    async def spend(
        self,
        *,
        agent_id: str,
        tokens: int,
        org_limit: int,
        agent_limit: int | None,
    ) -> SpendOutcome:
        if tokens <= 0:
            return SpendOutcome(ok=True)

        scope = agent_scope(agent_id)
        org_used = self._used.get(ORG_SCOPE, 0)
        seat_used = self._used.get(scope, 0)

        if org_limit and org_used + tokens > org_limit:
            return SpendOutcome(
                ok=False,
                rejected_scope="org",
                rejected_used=org_used,
                rejected_limit=org_limit,
            )
        if agent_limit and seat_used + tokens > agent_limit:
            return SpendOutcome(
                ok=False,
                rejected_scope="agent",
                rejected_used=seat_used,
                rejected_limit=agent_limit,
            )

        self._used[ORG_SCOPE] = org_used + tokens
        self._used[scope] = seat_used + tokens
        return SpendOutcome(
            ok=True,
            org_used=self._used[ORG_SCOPE],
            agent_used=self._used[scope],
        )

    async def usage(self, scope: str) -> int:
        return self._used.get(scope, 0)

    async def reset(self, scope: str = "") -> int:
        if scope:
            return 1 if self._used.pop(scope, None) is not None else 0
        count = len(self._used)
        self._used.clear()
        return count

File: internal/store/budgets.go
package store

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"strings"
	"time"
)

// The shared token counter.
//
// USAGE IS SHARED, CAPS ARE NOT. A cap belongs to a config epoch — a revision
// that raises a ceiling takes effect on the next turn — while the counter has
// to be one number across the fleet, because per-process counters mean N nodes
// each spend the whole allowance. So the limit travels IN on every call and
// the row holds only what has been spent.

// OrgScope is the company-wide counter's key.
const OrgScope = "org"

// AgentScope is one seat's counter key.
//
// Keyed on the DERIVED agent id rather than the handle, matching the diary and
// the episodes: renaming a handle then starts a fresh budget rather than
// inheriting the spend of whoever held the name before.
func AgentScope(agentID string) string { return "agent:" + agentID }

// Spend is what one charge did.
type Spend struct {
	// OK is false when a scope refused. RefusedScope, RefusedUsed and
	// RefusedLimit then say WHICH and by how much — "the company is out"
	// and "this seat is out" send an operator to different places, and a
	// bare refusal sends them to neither.
	OK           bool
	RefusedScope string
	RefusedUsed  int
	RefusedLimit int

	// OrgUsed and AgentUsed are the counters after a successful charge.
	OrgUsed   int
	AgentUsed int
}

// Budgets is the shared token counter.
type Budgets struct{ db *DB }

// Budgets returns the counter backed by this database.
func (d *DB) Budgets() *Budgets { return &Budgets{db: d} }

// bumpSQL is the atomic check-and-increment.
//
// The WHERE on DO UPDATE is what makes it atomic: a peer racing the last of a
// cap either lands inside it or updates nothing. Note it applies ONLY to the
// update branch — a scope with no row yet takes the INSERT, which has no
// existing value to test — which is why [Budgets.Spend] screens an
// over-cap charge before it gets here.
//
// No RETURNING: it is outside the intersection of the two certified drivers
// (rewrite/decisions/002), so the outcome is read from RowsAffected and the
// value from a following SELECT inside the same transaction.
const bumpSQL = `
INSERT INTO token_budget_usage (scope, used_tokens, updated_at)
VALUES (?, ?, ?)
ON CONFLICT (scope) DO UPDATE
SET used_tokens = token_budget_usage.used_tokens + excluded.used_tokens,
    updated_at  = excluded.updated_at
WHERE ? = 0 OR token_budget_usage.used_tokens + excluded.used_tokens <= ?`

// Charge checks and increments both scopes, atomically.
//
// ORG FIRST, THEN THE SEAT, in one transaction — so a seat that is refused
// unwinds the org bump it already made. Charging the org for a turn that never
// ran would let a company exhaust its budget on work it did not do.
//
// A limit of 0 is UNLIMITED, matching the config: `token_budget: 0` is how an
// operator says "no ceiling", and reading it as "no allowance" would stop every
// company that never set one.
//
// An error means the counter could not be reached, which is NOT a refusal and
// must not be reported as one: the caller fails the turn rather than telling an
// agent it is out of budget.
func (b *Budgets) Charge(ctx context.Context, agentID string, tokens, orgLimit, agentLimit int) (Spend, error) {
	if tokens <= 0 {
		// Not an error and not a charge. A phase whose provider reported
		// nothing still ran, and refusing it would stop a company over a
		// backend that omits usage.
		return Spend{OK: true}, nil
	}
	seat := AgentScope(agentID)

	// A charge larger than a whole cap can never fit, and the INSERT branch
	// of the upsert has no existing value to test against — so it is
	// screened here, before anything is written. Without this a first-ever
	// charge of a million tokens against a cap of ten would be accepted.
	for _, s := range []struct {
		name, scope string
		limit       int
	}{{"org", OrgScope, orgLimit}, {"agent", seat, agentLimit}} {
		if s.limit > 0 && tokens > s.limit {
			used, err := b.Used(ctx, s.scope)
			if err != nil {
				return Spend{}, err
			}
			return Spend{RefusedScope: s.name, RefusedUsed: used, RefusedLimit: s.limit}, nil
		}
	}

	var out Spend
	err := b.db.Tx(ctx, func(tx *sql.Tx) error {
		orgUsed, ok, err := bump(ctx, tx, OrgScope, tokens, orgLimit)
		if err != nil {
			return err
		}
		if !ok {
			out = Spend{RefusedScope: "org", RefusedUsed: orgUsed, RefusedLimit: orgLimit}
			return errRefused
		}
		agentUsed, ok, err := bump(ctx, tx, seat, tokens, agentLimit)
		if err != nil {
			return err
		}
		if !ok {
			out = Spend{RefusedScope: "agent", RefusedUsed: agentUsed, RefusedLimit: agentLimit}
			// Rolls the transaction back, which unwinds the org bump above.
			return errRefused
		}
		out = Spend{OK: true, OrgUsed: orgUsed, AgentUsed: agentUsed}
		return nil
	})
	if errors.Is(err, errRefused) {
		return out, nil
	}
	if err != nil {
		return Spend{}, fmt.Errorf("store: charge budget: %w", err)
	}
	return out, nil
}

// errRefused rolls a charge back without being an error to the caller. A
// sentinel rather than a bool out-parameter, because Tx's contract is that a
// non-nil error rolls back — which is exactly what a refusal needs.
var errRefused = errors.New("store: budget refused")

// bump applies one scope's charge, reporting the resulting usage and whether
// it fit.
func bump(ctx context.Context, tx *sql.Tx, scope string, tokens, limit int) (int, bool, error) {
	res, err := tx.ExecContext(ctx, bumpSQL, scope, tokens, EncodeTime(now()), limit, limit)
	if err != nil {
		return 0, false, err
	}
	rows, err := res.RowsAffected()
	if err != nil {
		return 0, false, err
	}
	used, err := usedIn(ctx, tx, scope)
	if err != nil {
		return 0, false, err
	}
	return used, rows > 0, nil
}

func usedIn(ctx context.Context, tx *sql.Tx, scope string) (int, error) {
	var used int
	err := tx.QueryRowContext(ctx,
		`SELECT used_tokens FROM token_budget_usage WHERE scope = ?`, scope).Scan(&used)
	if errors.Is(err, sql.ErrNoRows) {
		return 0, nil
	}
	return used, err
}

// Used reports one scope's spend. A scope with no row has spent nothing.
func (b *Budgets) Used(ctx context.Context, scope string) (int, error) {
	var used int
	err := b.db.sql.QueryRowContext(ctx,
		`SELECT used_tokens FROM token_budget_usage WHERE scope = ?`, scope).Scan(&used)
	if errors.Is(err, sql.ErrNoRows) {
		return 0, nil
	}
	if err != nil {
		return 0, fmt.Errorf("store: budget usage %q: %w", scope, err)
	}
	return used, nil
}

// Usage is one scope's counter, for the operator surface.
type Usage struct {
	Scope     string
	Used      int
	UpdatedAt time.Time
}

// List returns every counter, org first then seats by scope.
//
// Ordered so the operator surface does not have to sort, and so two reads of
// an unchanged table are byte-identical — a listing that reshuffled would make
// a diff of two captures unreadable.
func (b *Budgets) List(ctx context.Context) ([]Usage, error) {
	rows, err := b.db.sql.QueryContext(ctx,
		`SELECT scope, used_tokens, updated_at FROM token_budget_usage ORDER BY scope`)
	if err != nil {
		return nil, fmt.Errorf("store: list budgets: %w", err)
	}
	defer func() { _ = rows.Close() }()

	var out []Usage
	for rows.Next() {
		var (
			u  Usage
			at int64
		)
		if err := rows.Scan(&u.Scope, &u.Used, &at); err != nil {
			return nil, fmt.Errorf("store: list budgets: scan: %w", err)
		}
		u.UpdatedAt = DecodeTime(at)
		out = append(out, u)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("store: list budgets: %w", err)
	}
	// "org" sorts before "agent:…" alphabetically? It does not — so the
	// order is fixed here rather than left to the collation, which differs
	// between the two drivers for exactly this kind of key.
	for i, u := range out {
		if u.Scope == OrgScope && i != 0 {
			out[0], out[i] = out[i], out[0]
			break
		}
	}
	return out, nil
}

// Reset zeroes one scope, or every scope when given "".
//
// An operator action, never a schedule. A budget is a ceiling for the life of
// a deployment; a table that rolled itself over would silently re-arm a
// company somebody had stopped on purpose.
func (b *Budgets) Reset(ctx context.Context, scope string) (int64, error) {
	query := `DELETE FROM token_budget_usage`
	var args []any
	if strings.TrimSpace(scope) != "" {
		query += ` WHERE scope = ?`
		args = append(args, scope)
	}
	res, err := b.db.sql.ExecContext(ctx, query, args...)
	if err != nil {
		return 0, fmt.Errorf("store: reset budget: %w", err)
	}
	n, err := res.RowsAffected()
	if err != nil {
		return 0, fmt.Errorf("store: reset budget: %w", err)
	}
	return n, nil
}
